src/zenml/utils/notebook_utils.py

```diff
@@ -0,0 +1,249 @@
+                        "notebook that you're currently using.",
+                        ENV_ZENML_NOTEBOOK_PATH,
+                    )
+                    _ACTIVE_NOTEBOOK_PATH = None
+
+        if _ACTIVE_NOTEBOOK_PATH:
+            relative_path = os.path.relpath(
+                _ACTIVE_NOTEBOOK_PATH, source_utils.get_source_root()
+            )
+            if relative_path.startswith(os.path.pardir):
+                logger.warning(
+                    "Notebook path %s is not in source root %s, ignoring it..",
+                    _ACTIVE_NOTEBOOK_PATH,
+                    source_utils.get_source_root(),
+                )
+                _ACTIVE_NOTEBOOK_PATH = None
+
+    return _ACTIVE_NOTEBOOK_PATH
+
+
+def load_notebook(notebook_path: str) -> Dict[str, Any]:
+    """Load a notebook.
+
+    Args:
+        notebook_path: Path to the notebook.
+
+    Raises:
+        FileNotFoundError: If no notebook exist at the path.
+
+    Returns:
+        Dictionary of the notebook.
+    """
+    if not os.path.exists(notebook_path):
+        raise FileNotFoundError(
+            f"Notebook at path {notebook_path} does not exist."
+        )
+
+    with open(notebook_path) as f:
+        notebook_json = json.loads(f.read())
+
+    return notebook_json
+
+
+def is_running_in_notebook(notebook_path: str) -> None:
+    """Check if the current code is running in the given notebook.
+
+    Args:
+        notebook_path: Path of the notebook.
+
+    Returns:
+        If the current code is running in the given notebook.
+    """
+    if not Environment.in_notebook():
+        return False
+
+    try:
+        notebook_json = load_notebook(notebook_path)
+    except FileNotFoundError:
+        return False
+
+    active_cell_id = get_active_notebook_cell_id()
+
+    for cell in notebook_json["cells"]:
+        if cell["id"] == active_cell_id:
+            # This isn't a 100% guarantee if the notebook was copied but a good
+            # quick check
+            return True
+
+    return False
+
+
+def extract_notebook_cell_code(notebook_path: str, cell_id: str) -> str:
+    """Extract code from a notebook cell.
+
+    Args:
+        notebook_path: Path to the notebook file.
+        cell_id: ID of the cell for which to extract the code.
+
+    Raises:
+        RuntimeError: If the notebook contains no cell with the given ID.
+
+    Returns:
+        The cell content.
+    """
+    notebook_json = load_notebook(notebook_path)
+
+    for cell in notebook_json["cells"]:
+        if cell["id"] == cell_id:
+            if cell["cell_type"] != "code":
+                logger.warning(
+                    "Trying to extract code from notebook cell, but the cell "
+                    "type is %s. Continuing anyway..",
+                    cell["cell_type"],
+                )
+
+            # TODO: Should we exclude lines starting with special symbols here?
+            # Or whould that mess with some code?
```

Comment:
Special like magic cells, for instance?

Reply:
Yep exactly. I guess lines that start with `%` or `!`, but that could in theory remove some python lines right?

Reply:
I can do this
```python
%load_ext autoreload

import this
```
So I would be cautious with this. Maybe just throw out the cells with `%%` in the beginning?
But again:
```bash
%%sh

python3 -c "import this"
```

Can we just keep them?

Reply:
Yeah combined cells are tricky. How about something like this?

```
for line in cell:
  if line.startswith("%") or line.startswith("!"):
    logger.warning("Remove all non-python code from your cells")
```

Reply:
Should be good